tracking-vuln/README.md

```diff
@@ -0,0 +1,78 @@
+# Content provider tracking vulnerability
+
+## Brief theory
+We have an extension that spawns easter eggs on the websites you visit and lets you collect these eggs. This extension loads the images from a third-party webserver. Each time a new egg is to be spawned, an HTTP-request is sent to this third-party image provider. Since we spawn an egg each time the user loads a page, an HTTP request is sent each time the user loads a page.
+
+The problem with this extension is that when the HTTP requests are sent they contain not only the IP-address of the user but also the name of the website (more precisely [the domain, the host and the scheme](https://webmasters.stackexchange.com/questions/69477/how-to-understand-scheme-host-domainport-path-filename)) the request is sent from; set in the so-called HTTP "referer" header. This means that each time the user loads a website, its name and the IP-address of the user is sent to the third-party image provider. This allows the image provider to construct the browsing history of a user if it so wishes.
+
+## How to run
+
+### Web server
+As outlined in the [main README](https://github.com/andreaskth/securing-chrome-extensions/tree/main#how-to-run-a-node-web-server), enter the web server folder (here called 
+*image-provider-webserver*) and run `npm install`, and then run `node app.js` in the same folder to start the server. The default port used by the server is 3000; if you are using this port for something else you can change the port in `app.js` (don't forget to change the other occurrences of 3000 in this document as well).
+
+When the web server is running, you can visit `localhost:3000` to verify that you can see a list of links pointing to images.  
+
+To make the web server externally accessible, you can for example utilize *ngrok* as described in the [main README]((https://github.com/andreaskth/securing-chrome-extensions#how-to-make-web-server-externally-accessible-with-ngrok)).
+
+### Extension
+The vulnerable extensions is located in *vulnerable-easter-eggstention*. In the content script `egg.js` change the `imageSrc` variable on the first line to point to the URL where the web server is hosted (for example, the ngrok.io URL if you are using that). The addition of `"/images/"` on the second line should not be removed since this path is necessary to access the individual images.
```

Comment:
```suggestion
The vulnerable extension is located in *vulnerable-easter-eggstention*. In the content script `egg.js` change the `imageSrc` variable on the first line to point to the URL where the web server is hosted (for example, the ngrok.io URL if you are using that). The addition of `"/images/"` on the second line should not be removed since this path is necessary to access the individual images.
```